File: dataset/Vanilla/OPENAI_gpt-5-2025-08-07/Task189.py
import threading

# Prints number to console
def printNumber(x: int):
    print(x, end='')

class ZeroEvenOdd:
    def __init__(self, n: int):
        self.n = n
        self.zero_sem = threading.Semaphore(1)
        self.even_sem = threading.Semaphore(0)
        self.odd_sem = threading.Semaphore(0)

    def zero(self, printNumber):
        for i in range(1, self.n + 1):
            self.zero_sem.acquire()
            printNumber(0)
            if i % 2 == 1:
                self.odd_sem.release()
            else:
                self.even_sem.release()

    def even(self, printNumber):
        for i in range(2, self.n + 1, 2):
            self.even_sem.acquire()
            printNumber(i)
            self.zero_sem.release()

    def odd(self, printNumber):
        for i in range(1, self.n + 1, 2):
            self.odd_sem.acquire()
            printNumber(i)
            self.zero_sem.release()

def run_zero_even_odd(n: int) -> str:
    zeo = ZeroEvenOdd(n)
    out = []
    lock = threading.Lock()

    def p(x: int):
        with lock:
            out.append(str(x))
        printNumber(x)

    ta = threading.Thread(target=zeo.zero, args=(p,))
    tb = threading.Thread(target=zeo.even, args=(p,))
    tc = threading.Thread(target=zeo.odd, args=(p,))
    ta.start(); tb.start(); tc.start()
    ta.join(); tb.join(); tc.join()
    return ''.join(out)

if __name__ == "__main__":
    tests = [1, 2, 5, 6, 7]
    for n in tests:
        run_zero_even_odd(n)
        print()

File: dataset/Vanilla/OPENAI_gpt-5-2025-08-07/Task189.java
import java.util.concurrent.Semaphore;
import java.util.function.IntConsumer;

public class Task189 {

    // Prints number to console
    public static void printNumber(int x) {
        System.out.print(x);
    }

    static class ZeroEvenOdd {
        private int n;
        private final Semaphore zeroSem = new Semaphore(1);
        private final Semaphore evenSem = new Semaphore(0);
        private final Semaphore oddSem = new Semaphore(0);

        public ZeroEvenOdd(int n) {
            this.n = n;
        }

        public void zero(IntConsumer printNumber) {
            try {
                for (int i = 1; i <= n; i++) {
                    zeroSem.acquire();
                    printNumber.accept(0);
                    if ((i & 1) == 1) {
                        oddSem.release();
                    } else {
                        evenSem.release();
                    }
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }

        public void even(IntConsumer printNumber) {
            try {
                for (int i = 2; i <= n; i += 2) {
                    evenSem.acquire();
                    printNumber.accept(i);
                    zeroSem.release();
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }

        public void odd(IntConsumer printNumber) {
            try {
                for (int i = 1; i <= n; i += 2) {
                    oddSem.acquire();
                    printNumber.accept(i);
                    zeroSem.release();
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
    }

    // Runs the ZeroEvenOdd with threads; returns the produced string (also prints to console)
    public static String runZeroEvenOdd(int n) {
        ZeroEvenOdd zeo = new ZeroEvenOdd(n);
        StringBuilder sb = new StringBuilder();
        IntConsumer p = x -> {
            synchronized (sb) { sb.append(x); }
            printNumber(x);
        };

        Thread tA = new Thread(() -> zeo.zero(p));
        Thread tB = new Thread(() -> zeo.even(p));
        Thread tC = new Thread(() -> zeo.odd(p));
        tA.start();
        tB.start();
        tC.start();
        try {
            tA.join();
            tB.join();
            tC.join();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        return sb.toString();
    }

    public static void main(String[] args) {
        int[] tests = {1, 2, 5, 6, 7};
        for (int n : tests) {
            runZeroEvenOdd(n);
            System.out.println();
        }
    }
}